File: core/src/main/java/pl/dybuk/health/core/common/rx/SimpleFuture.java
package pl.dybuk.health.core.common.rx;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class SimpleFuture<T> implements Future<T> {

    private T result;

    private Throwable error;

    private boolean isDone = false;
    private boolean isCanceled = false;

    @Override
    public synchronized boolean cancel(boolean mayInterruptIfRunning) {
        isCanceled = true;
        return true;
    }

    @Override
    public synchronized boolean isCancelled() {
        return isCanceled;
    }

    @Override
    public synchronized boolean isDone() {
        return isDone;
    }

    @Override
    public synchronized T get() throws ExecutionException, InterruptedException {
        while (!isDone) {
            wait();
        }

        if (error != null) {
            throw new ExecutionException(error);
        }

        return result;
    }

    @Override
    public synchronized T get(long timeout, TimeUnit unit) throws ExecutionException, InterruptedException, TimeoutException {
        if (!isDone) {
            wait(unit.toMillis(timeout));
        }

        if (!isDone) {
            throw new TimeoutException();
        }

        if (error != null) {
            throw new ExecutionException(error);
        }

        return result;
    }

    public synchronized void setResult(T result) {
        this.isDone = true;
        this.result = result;
        this.error = null;
        notifyAll();
    }

    public synchronized void setError(Throwable error) {
        this.isDone = true;
        this.result = null;
        this.error = error;
        notifyAll();
    }
}

File: core/src/test/java/pl/dybuk/health/core/common/rx/SimpleFuture.kt
package pl.dybuk.health.core.common.rx

import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.runners.MockitoJUnitRunner
import pl.dybuk.health.core.Dummy
import pl.dybuk.health.core.chat.value.Id
import org.fest.assertions.api.Assertions.assertThat
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

@RunWith(MockitoJUnitRunner::class)
class SimpleFutureTest {

    @Test(expected = TimeoutException::class)
    fun testTimeout() {
        val future = SimpleFuture<Dummy>()
        future.get(100, TimeUnit.MILLISECONDS)
    }

    @Test()
    fun testValueAlreadyPresentGetWithTime() {
        // When
        val future = SimpleFuture<Dummy>()

        // Given
        future.setResult(Dummy(Id(1)))

        // Then
        val result = future.get(100, TimeUnit.MILLISECONDS)
        assertThat(result).isEqualTo(Dummy(Id(1)))
    }

    @Test()
    fun testValueAlreadyPresentSimpleGet() {
        // When
        val future = SimpleFuture<Dummy>()

        // Given
        future.setResult(Dummy(Id(1)))

        // Then
        assertThat(future.get()).isEqualTo(Dummy(Id(1)));
    }


    @Test()
    fun testValuePushGetWithTime() {
        // When
        val future = SimpleFuture<Dummy>()
        Thread { Thread.sleep(300);  future.setResult(Dummy(Id(1))) }.start()

        // Given
        val result = future.get(5000, TimeUnit.MILLISECONDS)

        // Then
        assertThat(result).isEqualTo(Dummy(Id(1)))
    }

    @Test()
    fun testValuePushSimpleGet() {
        // When
        val future = SimpleFuture<Dummy>()
        Thread { Thread.sleep(300); future.setResult(Dummy(Id(1))) }.start()

        // Given
        val result = future.get()

        // Then
        assertThat(result).isEqualTo(Dummy(Id(1)))
    }



    @Test(expected = ExecutionException::class)
    fun testErrorAlreadyPresentGetWithTime() {
        // When
        val future = SimpleFuture<Dummy>()

        // Given
        future.setError(Throwable("ERROR"))

        // Then
        future.get(100, TimeUnit.MILLISECONDS)
    }

    @Test(expected = ExecutionException::class)
    fun testErrorAlreadyPresentSimpleGet() {
        // When
        val future = SimpleFuture<Dummy>()

        // Given
        future.setError(Throwable("ERROR"))

        // Then
        future.get()
    }


    @Test(expected = ExecutionException::class)
    fun testErrorPushGetWithTime() {
        // When
        val future = SimpleFuture<Dummy>()
        Thread { Thread.sleep(300); future.setError(Throwable("ERROR")) }.start()

        // Given
        val result = future.get(5000, TimeUnit.MILLISECONDS)

        // Then
        assertThat(result).isEqualTo(Dummy(Id(1)))
    }

    @Test(expected = ExecutionException::class)
    fun testErrorPushSimpleGet() {
        // When
        val future = SimpleFuture<Dummy>()
        Thread { Thread.sleep(300); future.setError(Throwable("ERROR")) }.start()

        // Given
        val result = future.get()

        // Then
        assertThat(result).isEqualTo(Dummy(Id(1)))
    }

    @Test
    fun testStateChangeToCancel() {
        // When
        val future = SimpleFuture<Dummy>()
        assertThat(future.isCancelled).isFalse
        assertThat(future.isDone).isFalse

        // Given
        future.cancel(true)

        // Then
        assertThat(future.isCancelled).isTrue
        assertThat(future.isDone).isFalse
    }

    @Test
    fun testStateChangeOnResultSet() {
        // When
        val future = SimpleFuture<Dummy>()
        assertThat(future.isCancelled).isFalse
        assertThat(future.isDone).isFalse

        // Given
        future.setResult(Dummy(Id(1)))

        // Then
        assertThat(future.isCancelled).isFalse
        assertThat(future.isDone).isTrue
    }

    @Test
    fun testStateChangeOnErrorSet() {
        // When
        val future = SimpleFuture<Dummy>()
        assertThat(future.isCancelled).isFalse
        assertThat(future.isDone).isFalse

        // Given
        future.setError(Throwable("ERROR"))

        // Then
        assertThat(future.isCancelled).isFalse
        assertThat(future.isDone).isTrue
    }

}